lib/Twig/Environment.php

```diff
@@ -590,7 +591,10 @@ public function getCharset()
      */
     public function hasExtension($class)
     {
-        return $this->extensionSet->hasExtension($class);
+        if( !isset( $this->extensionCache[$this->optionsHash][$class] ) ) {
+            return $this->extensionSet->hasExtension($class);
+        }
+        return $this->extensionCache[$this->optionsHash][$class];
```

Comment:
You do not write to `extensionCache` in `hasExtension`, so it does improve the performance when calling `hasExtension` multiple times.